src/gt4py/next/otf/compiled_program.py

```diff
@@ -0,0 +1,275 @@
+    will error. In the latter case, the pool needs to be filled with call(s)
+    to 'compile' before it can be used.
+    """
+
+    backend: gtx_backend.Backend
+    definition_stage: ffront_stages.ProgramDefinition
+    program_type: ts_ffront.ProgramType
+    static_params: Sequence[str] | None = None  # not ordered
+
+    # TODO(havogt): This dict could be replaced by a `functools.cache`d method
+    # and appropriate hashing of the arguments.
+    _compiled_programs: dict[
+        _CompiledProgramsKey,
+        stages.CompiledProgram | concurrent.futures.Future[stages.CompiledProgram],
+    ] = dataclasses.field(default_factory=dict)
+
+    def __postinit__(self) -> None:
+        # TODO(havogt): We currently don't support pos_only or kw_only args at the program level.
+        # This check makes sure we don't miss updating this code if we add support for them in the future.
+        assert not self.program_type.definition.kw_only_args
+        assert not self.program_type.definition.pos_only_args
+
+    def __call__(
+        self, *args: Any, offset_provider: common.OffsetProvider, enable_jit: bool, **kwargs: Any
+    ) -> None:
+        """
+        Calls a program with the given arguments and offset provider.
+
+        If the program is not in cache, it will jit compile with static arguments
+        (defined by 'static_params') in case `enable_jit` is True. Otherwise,
+        it is an error.
+        """
+        args, kwargs = type_info.canonicalize_arguments(self.program_type, args, kwargs)
+        offset_provider_type = _offset_provider_to_type_unsafe(offset_provider)
+        key = _CompiledProgramsKey(
+            tuple(args[i] for i in self._static_arg_indices),
+            offset_provider_type,
+        )
+        try:
+            self._compiled_programs[key](*args, **kwargs, offset_provider=offset_provider)  # type: ignore[operator] # for performance: try to call first...
+        except TypeError:  # 'Future' object is not callable
+            # ... otherwise we resolve the future and call again
+            program = self._resolve_future(key)
+            program(*args, **kwargs, offset_provider=offset_provider)
+        except KeyError as e:
+            if enable_jit:
+                assert self.static_params is not None
+                static_args = {name: args[self._param_index(name)] for name in self.static_params}
+                self._compile_variant(static_args=static_args, offset_provider=offset_provider)
+                return self(
+                    *args, offset_provider=offset_provider, enable_jit=False, **kwargs
+                )  # passing `enable_jit=False` because a cache miss should be a hard-error in this call`
+            raise RuntimeError("No program compiled for this set of static arguments.") from e
+
+    def _param_index(self, name: str) -> int:
+        return list(self.program_type.definition.pos_or_kw_args.keys()).index(name)
```

Comment:
Shouldn't this be converted to a `_param_indices` cached property? There is no need to reconstruct the list in every call
```suggestion
    @functools.cached_property
    def _param_index(self, name: str) -> list[int]:
        return list(self.program_type.definition.pos_or_kw_args.keys())
```

Reply:
This concrete suggestion is wrong but the handling of the static arg indices and params has been slightly enhanced in a different way that also avoids recomputing the list multiple times.